mmtk/src/gc_trigger.rs

```diff
@@ -0,0 +1,330 @@
+        self.gc_end_time
+            .store(gc_end_time as usize, Ordering::Relaxed);
+
+        let reserved_pages_now =
+            mmtk.get_plan().get_reserved_pages() + self.pending_pages.load(Ordering::SeqCst);
+        let heap_size = conversions::pages_to_bytes(reserved_pages_now);
+
+        let user_max = self.max_total_memory.load(Ordering::Relaxed) * 80 / 100;
+        let alloc_diff = self.before_free_heap_size.load(Ordering::Relaxed)
+            - self.old_heap_size.load(Ordering::Relaxed);
+        let freed_diff = self.before_free_heap_size.load(Ordering::Relaxed) - heap_size;
+        self.old_heap_size.store(heap_size, Ordering::Relaxed);
+
+        // update the heap target only if the user did not force a GC
+        let gc_auto = !mmtk.is_user_triggered_collection();
+        if gc_auto {
+            let mut target_allocs: f64 = 0.0;
+            const ALLOC_SMOOTH_FACTOR: f64 = 0.95;
+            const COLLECT_SMOOTH_FACTOR: f64 = 0.5;
+            const TUNING_FACTOR: f64 = 2e4;
+            let alloc_mem = mmtk_jl_gc_smooth(
+                self.old_alloc_diff.load(Ordering::Relaxed),
+                alloc_diff,
+                ALLOC_SMOOTH_FACTOR,
+            );
+            let alloc_time = mmtk_jl_gc_smooth(
+                self.old_mut_time.load(Ordering::Relaxed),
+                self.mutator_time.load(Ordering::Relaxed),
+                ALLOC_SMOOTH_FACTOR,
+            );
+            let gc_mem = mmtk_jl_gc_smooth(
+                self.old_freed_diff.load(Ordering::Relaxed),
+                freed_diff,
+                COLLECT_SMOOTH_FACTOR,
+            );
+            let gc_time = mmtk_jl_gc_smooth(
+                self.old_pause_time.load(Ordering::Relaxed),
+                pause as usize, // note that we the stock GC discounts the sweep time
+                COLLECT_SMOOTH_FACTOR,
+            );
+
+            self.old_alloc_diff.store(alloc_mem, Ordering::Relaxed);
+            self.old_mut_time.store(alloc_time, Ordering::Relaxed);
+            self.old_freed_diff.store(gc_mem, Ordering::Relaxed);
+            self.old_pause_time.store(gc_time, Ordering::Relaxed);
+
+            // thrashing estimator: if GC time more than 50% of the runtime
+            if pause > self.mutator_time.load(Ordering::Relaxed)
+                && self.thrash_counter.load(Ordering::Relaxed) >= 4
+            {
+                self.thrash_counter.fetch_add(1, Ordering::Relaxed);
+            } else if self.thrash_counter.load(Ordering::Relaxed) > 0 {
+                self.thrash_counter.fetch_sub(1, Ordering::Relaxed);
+            }
+
+            if alloc_mem != 0 && alloc_time != 0 && gc_mem != 0 && gc_time != 0 {
+                let alloc_rate = alloc_mem as f64 / alloc_time as f64;
+                let gc_rate = gc_mem as f64 / gc_time as f64;
+                target_allocs = (heap_size as f64 * alloc_rate / gc_rate).sqrt() * TUNING_FACTOR;
```

Comment:
The computation looks like mem balancer. If this is Julia's current heuristics, I assume Julia is using a variant of mem balancer (e.g. the thrashing detection).

Reply:
That's right. I think it does, and that's what I wanted to say in the MMTk meeting. 
However, looking at [`TUNING_FACTOR`](https://github.com/JuliaLang/julia/blob/090e291eba1b7668bfc5234215750b4bd80f5387/src/gc-stock.c#L3225) from Julia, it looks much different than what [mmtk-core](https://github.com/mmtk/mmtk-core/blob/051bc7470feef915c445305301e6113f86d3957b/src/util/heap/gc_trigger.rs#L569) implements? But again, the equation they use seems different...

Reply:
We did not 'tune' for mmtk's mem balancer. They multiply with the `TUNING_FACTOR`, we divide by the `TUNING_FACTOR`. But set this aside, it looks like that they are using a variant of mem balancer.

Reply:
Yes, our heuristics started with stock membalancer and then we changed the quite a bit, notably by allowing it to scale a lot faster when small